<p align="center">
  <img src="assets/William-Shakespeare.png" style="width: 350px"/>
</p>

# Introducción a la Ciencia de Datos: Laboratorio 1

William Shakespeare, nació el 23 de abril de 1564 en Stratford-upon-Avon, Inglaterra y en lo que hoy en día parecería una vida corta (52 años), se transformó en una figura titánica del mundo de la literatura. Este dramaturgo y poeta inglés dejó un legado imborrable con sus más de 39 obras literarias, existen al menos dos corrientes que discuten incluso hoy en día la atribución de ciertas obras, entre entre las que se destacan sus tragedias y comedias, obras como "Hamlet", "Romeo y Julieta" y "El rey Lear". Ya sea si has leído alguna obra de William Shakespeare o no, es muy probable que reconozcas algunas frases con origen en su obra como "Ser o no ser, esa es la cuestión" o "El amor es un humo hecho con el vapor de suspiros". Estas líneas no solo demuestran su maestría lingüística, sino que también reflejan las intrigas universales sobre el amor, el poder y la tragedia, manteniendo su relevancia a través de los siglos.

En este trabajo llevado adelante en el contexto del primer Laboratorio [2] del curso Introducción a la Ciencia de Datos de la Facultad de Ingeniería, UdelaR, edición 2024, nos proponemos adentrarnos en la obra de William Shakespeare con un enfoque de ciencia de datos, analizando sus principales obras utilizando algunas técnicas sencillas de análisis de datos.

Esperamos que disfrutes este viaje a través de los datos, el tiempo y principalmente, de la lengua inglesa, tanto como nosotros lo hemos disfrutado.


_"Ten más de lo que muestras habla menos de lo que sabes."_

_William Shakespeare_

## Índice
<a name="index"></a>

1. [Imports & Utils](#imports)
2. [Adquisición de Datos](#data-adquisition)
3. [Entendimiento de los Datos](#data-understanding)
    1. [Dominio del Problema](#domain)
    2. [EDA: Análisis Exploratorio de Datos](#eda)
4. [Procesamiento de los Datos](#data-processing)
5. [Análisis de Datos](#data-analysis)
6. [Conclusiones](#conclusions)
7. [Referencias](#references)

## 1. Imports & Utils <a name="imports"></a>
[Volver al Inicio](#index)

Esta sección contiene todos los imports de dependencias y librerias utilizadas por este proyecto. También contiene la definición de funciones auxiliares utilzadas para obtener los datos y procesarlos. Por último, recuerde instalar los requerimientos (`requirements.txt`) en el mismo entorno donde está ejecutando este notebook y de esa forma evitar errores de import de dependencias (ver [README](README.md)).

In [1]:
import os
from pathlib import Path
from time import time
from typing import Any, List, Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import spacy
import tqdm
from sqlalchemy import create_engine
from wordcloud import WordCloud
from ydata_profiling import ProfileReport

A continuación definimos algunos parámetros globales del notebook como rutas por defecto y otras configuraciones para centralizar la configuración de experimentos.

In [2]:
# Globals definitions

DATA_FOLDER = os.path.join(
    "data", "shakespeare"
)  # Path en donde se almacenan los datos de laboratorio 1 en formato
DATA_SOURCE = "local"  # valid values: local | web
SHAKESPEARE_DB_CONN = (
    "mysql+pymysql://guest:relational@db.relational-data.org:3306/Shakespeare"
)
FIGURES_FOLDER = os.path.join("assets", "snapshoots")
DATA_REPORTS = os.path.join("assets", "reports")
DEFAULT_TOP_ROWS_DISPLAY = 10  # Por default cuantas row mostrar con TOP

Las siguientes funciones fueron definidas por el equipo docente y provistas como parte de los recursos del Laboratorio 1.

In [4]:
# Definidas por el equipo docente


def load_table(data_dir, table_name, engine):
    """
    Leer la tabla con SQL y guardarla como CSV,
    o cargarla desde el CSV si ya existe
    """
    path_table = data_dir / f"{table_name}.csv"
    if not path_table.exists():
        print(f"Consultando tabla con SQL: {table_name}")
        t0 = time()
        with engine.connect() as conn:
            df_table = pd.read_sql(
                sql=f"SELECT * FROM {table_name}", con=conn.connection
            )
        # df_table = pd.read_sql(f"SELECT * FROM {table_name}", engine)
        t1 = time()
        print(f"Tiempo: {t1 - t0:.1f} segundos")

        print(f"Guardando: {path_table}\n")
        df_table.to_csv(path_table)
    else:
        print(f"Cargando tabla desde CSV: {path_table}")
        df_table = pd.read_csv(path_table, index_col=[0])
    return df_table


def clean_text(df, column_name):
    result = df[column_name].str.lower()  # Convertir todo a minúsculas
    result = result.str.strip()  # Remueve espacios en blanco

    # Quitar signos de puntuación y cambiarlos por espacios (" ")
    # TODO: completar signos de puntuación faltantes
    for punc in ["[", "\n", ",", ":", ";", ".", "]", "(", ")", "?", "!", "'", "-", "\"", "{", "}"]:
        result = result.str.replace(punc, " ")
    return result

Las siguientes funciones auxiliares fueron definidas por nosotros para facilitar el análisis de datos de este laboratorio.

In [5]:
# Definidas por nosotros


def read_from_csv(path: str) -> pd.DataFrame:
    """Método para leer datos desde un archivo CSV local.

    Args:
        path (str): Ruta al archivo CSV

    Returns:
        pd.DataFrame: Dataframe con datos.
    """
    return pd.read_csv(path, sep=",", index_col=0)


def count_empty_values(df: pd.DataFrame) -> pd.Series:
    """Cuenta valores vacíos.

    Esta función cuenta valores vaciós en un dataframe en función del tipo de columna (object, int)
    utilizando ciertas convenciones para valores vaciós como que un np.nan y -1 ambos pueden ser valores
    válidos para representar un valor faltante o vacío en una columna nunérica.
    Args:
        df (pd.DataFrame): input dataframe

    Returns:
        pd.Series: conteo de vacíos
    """

    def is_empty(column):
        if column.dtype == object:  # Assuming object dtype for strings
            return column.isin([None, "", np.nan])
        elif column.dtype == int:
            return column.isin([None, np.nan, -1])

    empty_counts = df.apply(is_empty).sum()
    return empty_counts

def generate_pandas_report(df: pd.DataFrame, name: str, path:str) -> None:
    """Genera EDA report.

    Utiliando ydata-profiling genera un reporte exploratorio de los datos recibidos.
    Args:
        df (pd.DataFrame): Data Frame a analizar.
        name (str): Nombre del data frame.
    """

    profile = ProfileReport(df, title=f"Profiling Report {name}")
    profile.to_file(os.path.join(path, f"{name}_report.html"))

## 2. Adquisición de los Datos  <a name="data-adquisition"></a>
[Volver al Inicio](#index)

Las siguientes celdas se encarga de obtener los datos del Laboratorio 1 y cargarlos en dataframes de pandas para facilitar su análisis.

In [6]:
def load_dataframes() -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    """Descarga de datos.

    Este método se encarga de descargar los datos desde el repositorio público de Shakespeare por primera vez,
    guardando los datos de cada tabla en un archivo CSV separado.

    Returns:
        Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]: Dataframes con los datos de las tablas.
    """

    # Creamos el directorio DATA_FOLDER donde se guardarán los CSV
    data_dir = Path(DATA_FOLDER)
    data_dir.mkdir(parents=True, exist_ok=True)

    # Evitamos crear la conexión a la BD si vamos a trabajar local
    if DATA_SOURCE == "web":
        print(f"Creando conexión a la base usando url={SHAKESPEARE_DB_CONN}...")
        engine = create_engine(SHAKESPEARE_DB_CONN)
    elif DATA_SOURCE == "local":
        print("Evitando crear conexión a BD...")
        engine = None
    else:
        raise Exception(
            "Debe especificar un tipo de source válido para los datos: 'web' | 'local'."
        )

    # DataFrame con todas las obras:
    df_works = load_table(data_dir, "works", engine)

    # Todos los párrafos de todas las obras
    df_paragraphs = load_table(data_dir, "paragraphs", engine)

    # TODO: cargar el resto de las tablas
    # Completamos el código originalmente provisto por los docentes.

    # DataFrame con los chapters
    df_chapters = load_table(data_dir, "chapters", engine)

    # DataFrame con los chapters
    df_characters = load_table(data_dir, "characters", engine)

    return df_works, df_paragraphs, df_chapters, df_characters

En la siguiente celda cargamos los datos de las tablas: Works, Paragraphs, Chapters y Characters. Más adelante en la siguiente sección entramos en detalles sobre que son los datos de cada una de estas tablas. 

Por otro lado, los datos se encuentran disponibles en la web en [1] y es posible descargarlos mediante el método provisto por los docentes ```load_table()``` y la librería [SQLAlchemy](https://www.sqlalchemy.org/). Para simplificar este proceso implementamos el método ```load_dataframes()``` que se encarga de articular la descarga utilizando las herramientas mencionadas anteriormente. No obstante, no tiene sentido descargarse desde la web los datos, cada veze que se ejecuta este notebook. Por tal razón los datos se guardan localmente en archivos ```.csv``` en el directorio definido por ```DATA_FOLDER```. Notar que el método ```load_table()``` prevee esto mismo en caso de encontrar en el directorio destino un archivo .csv con el nombre de la tabla. Por otro lado, evitamos crear la conexión a la BD en caso que se quiera trabajar local, para evitar por ejemplo errores de conexión con la BD. Para esto usamos la variable DATA_SOURCE que nos indica el modo de trabajo:

* DATA_SOURCE='web' -> Crea conexión a la BD y utiliza ```load_dataframes()``` para descargar los datos.
* DATA_SOURCE='local' -> No crea la conexión a la BD y utiliza ```load_dataframes()``` para leer los datos localmente ya que ```load_table()``` encontrará los archivos .csv.

En la siguiente celda, se cargan los datos en los dataframes de nombre ```df_works, df_paragraphs, df_chapters, df_characters```.



In [7]:
print("Cargando los datos...")
df_works, df_paragraphs, df_chapters, df_characters = load_dataframes()

print(f"Works: {df_works.shape}")
print(f"Paragraphs: {df_paragraphs.shape}")
print(f"Chapters: {df_chapters.shape}")
print(f"Characters: {df_characters.shape}")
print("Datos cargados exitosamente!")

Cargando los datos...
Evitando crear conexión a BD...
Cargando tabla desde CSV: data/shakespeare/works.csv
Cargando tabla desde CSV: data/shakespeare/paragraphs.csv
Cargando tabla desde CSV: data/shakespeare/chapters.csv
Cargando tabla desde CSV: data/shakespeare/characters.csv
Works: (43, 5)
Paragraphs: (35465, 5)
Chapters: (945, 5)
Characters: (1266, 4)
Datos cargados exitosamente!


## 3. Entendimiento de los Datos <a name="data-understanding"></a>
[Volver al Inicio](#index)

### 3.1. Dominio del Problema  <a name="domain"></a>


Más información acerca de las tablas disponibles en la base de datos [aquí](https://relational-data.org/dataset/Shakespeare). 

![img](assets/image_01.png)

### 3.2. EDA: Análisis Exploratorio de Datos  <a name="eda"></a>

Existen muchas técnicas para llevar adelante un análisis exploratorio de datos, algunas dentro de la intuición y otras de una naturaleza más estadísitca. En este trabajo, nos vamos a limitar a ejecutar algunos análisis básicos e intuitivos para principalmente ganar mayor conocimiento sobre los datos analizados y validar algunas hipótesis sobre la calidad de los datos y a su vez nos vamos a apoyar en la libreria [ydata-profiling](https://docs.profiling.ydata.ai/latest/) para realizar automáticamente un análisis más estadístico completo sobre los datos. 

Para profundizar sobre los objetivos de un análisis exploratorio de datos (EDA) y herramientas disponibles recomendamos la lectura de [4] y [5].

**Sobre el Análisis**:

Para cada tablas/datarame vamos a conducir principalmente los mismos análisis:

1. Vistaso rápido de los datos
2. Revisión de Tipos
3. Caracterísitcas Macro
4. Conteo de missing-values
5. Conteo de duplicados
6. Revisión de valores inválidos

#### 3.2.1. Works

Hechemos un vistaso rápido a los datos observando las primeras N filas. Para eso usamos la función [head()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) de pandas. Recordemos que las obras las tenemos cargadas en ```df_works```.

In [8]:
# Muestra TOP DEFAULT_TOP_ROWS_DISPLAY filas del dataframe
df_works.head(DEFAULT_TOP_ROWS_DISPLAY)

id                      Title                                  LongTitle  \
0   1              Twelfth Night            Twelfth Night, Or What You Will   
1   2  All's Well That Ends Well                  All's Well That Ends Well   
2   3       Antony and Cleopatra                       Antony and Cleopatra   
3   4             As You Like It                             As You Like It   
4   5           Comedy of Errors                       The Comedy of Errors   
5   6                 Coriolanus                                 Coriolanus   
6   7                  Cymbeline                 Cymbeline, King of Britain   
7   8                     Hamlet  Tragedy of Hamlet, Prince of Denmark, The   
8   9           Henry IV, Part I                History of Henry IV, Part I   
9  10          Henry IV, Part II               History of Henry IV, Part II   

   Date GenreType  
0  1599    Comedy  
1  1602    Comedy  
2  1606   Tragedy  
3  1599    Comedy  
4  1589    Comedy  
5  1607   Tragedy  
6  1609   History  
7  1600   Tragedy  
8  1597   History  
9  1597   History

Veamos que tipos infirió automáticamente pandas para cada columna, para asegurarnos que son los correctos y revisar si tenemos que hacer algún tipo de procesamiento previo. Para esto accedemos a la propiedad [.dypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html) del dataframe.

In [9]:
# Listado de tipos para las columnas
df_works.dtypes

id            int64
Title        object
LongTitle    object
Date          int64
GenreType    object
dtype: object

También podemos inferir propiedades interesantes del dataframe como los valores mínimo/maximo/avg de cada columna utilizando la función [describe()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) de pandas. Notar que estos valores existen solamente para columnas numéricas y no siempre es relevante su uso. En este caso por ejemplo, nos permite entender más sobre el período en el que se encuentran las obras de Shakespeare analizadas el cual va desde el año 1589 hasta el año 1612. Por otro lado la media de las obras se encuentra hacia el año 1599.

Por otro lado, podemos observar que estamos analizando 43 obras. Por último, de acuerdo con Wikipedia [6], Shakespeare nació en el año 1564 y murió en el año 1616 y todos los valores de ```Date``` para works se encuentran dentro de dicho rango. Es un buen indicio de la calidad de los datos en esta columna.

In [10]:
# Valores descriptivos de cada columna
df_works.describe()

id         Date
count  43.000000    43.000000
mean   22.000000  1599.627907
std    12.556539     6.473474
min     1.000000  1589.000000
25%    11.500000  1594.000000
50%    22.000000  1599.000000
75%    32.500000  1605.000000
max    43.000000  1612.000000

Para hacer más sencillo el conteo de valores faltantes implementamos el método ```count_empty_values()``` que recibe como parámetro un dataframe y revisa todas las columnas del mismo, contando valores faltantes (missing values). Como sabemos, un valor faltante puede ser tanto un valor ```None``` como un ```0```, ```-1```, ```NaN```, ```''``` (string vacío). Depende del tipo de la columna y el problema principalmente. 

In [11]:
# Conteo de missing values por columna
empty_values = count_empty_values(df=df_works)
empty_values

id           0
Title        0
LongTitle    0
Date         0
GenreType    0
dtype: int64

Otro problema de calidad puede ser la existencia de valores repetídos, en cuyo caso aveces nos fuerza a tener que eliminarlos previo a realizar un análisis. Para esto vamos a utilizar la función [duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) de pandas. Notar que en algunas columnas no solo tenemos valores repetidos sino que tiene mucho sentído y además nos va a dar información relevante sobre la obra de Shakespeare.

In [12]:
# Revisamos ocurrencias duplicadas en cada columna del dataframe
duplicate_counts = {col: df_works[col].duplicated().sum() for col in df_works.columns}
duplicate_counts

{'id': 0, 'Title': 0, 'LongTitle': 0, 'Date': 20, 'GenreType': 38}

In [13]:
# Veamos cuales son los géneros que comprenden a la obra de Shakespeare
genres = df_works["GenreType"].unique().tolist()
print(
    f"La obra de Shakespeare se concentra en {len(genres)} géneros: {
        ', '.join(genres)}"
)

La obra de Shakespeare se concentra en 5 géneros: Comedy, Tragedy, History, Poem, Sonnet


No parece haber datos duplicados que requieran de un pre-procesamiento:

* ✅ No hay repetidos en columna ```id```
* ✅ No hay repetidos en columna ```Title```
* ✅ No hay repetidos en columna ```LongTitle```

Además:

* Hay varios años en los que Shakespeare publicó más de una obra (más adelante analizaremos esto en detalle)
* Hay repetición en los géneros lo cual tiene mucho sentido ya que son un grupo de apenas **5** géneros: _Comedy, Tragedy, History, Poem, Sonnet_


In [ ]:
# Valores inválidos
# NO TIENE

In [ ]:
# Genera ydata profiling report
generate_pandas_report(df_works, "works", DATA_REPORTS)

#### 3.2.2. Chapters

Repitamos el mismo procedimiento ahora utilizando el dataframe ```df_chapters``` que contiene los datos de capítulos de las obras.

In [16]:
# Muestra TOP DEFAULT_TOP_ROWS_DISPLAY filas del dataframe
df_chapters.head(DEFAULT_TOP_ROWS_DISPLAY)

id  Act  Scene            Description  work_id
0  18704    1      1  DUKE ORSINO’s palace.        1
1  18705    1      2         The sea-coast.        1
2  18706    1      3        OLIVIA’S house.        1
3  18707    1      4  DUKE ORSINO’s palace.        1
4  18708    1      5        OLIVIA’S house.        1
5  18709    2      1         The sea-coast.        1
6  18710    2      2              A street.        1
7  18711    2      3        OLIVIA’s house.        1
8  18712    2      4  DUKE ORSINO’s palace.        1
9  18713    2      5       OLIVIA’s garden.        1

In [17]:
# Listado de tipos para las columnas
df_chapters.dtypes

id              int64
Act             int64
Scene           int64
Description    object
work_id         int64
dtype: object

In [18]:
# Valores descriptivos de cada columna
df_chapters.describe()

id         Act       Scene     work_id
count    945.000000  945.000000  945.000000  945.000000
mean   19176.000000    2.637037   15.347090   22.063492
std      272.942302    1.485988   32.882111   12.683993
min    18704.000000    0.000000    0.000000    1.000000
25%    18940.000000    1.000000    2.000000   11.000000
50%    19176.000000    3.000000    3.000000   23.000000
75%    19412.000000    4.000000    6.000000   35.000000
max    19648.000000    5.000000  154.000000   43.000000

In [19]:
# Conteo de missing values
empty_values = count_empty_values(df=df_chapters)
empty_values

id             0
Act            0
Scene          0
Description    0
work_id        0
dtype: int64

In [20]:
# Revisamos ocurrencias duplicadas en cada columna del dataframe
duplicate_counts = {
    col: df_chapters[col].duplicated().sum() for col in df_chapters.columns
}
duplicate_counts

{'id': 0, 'Act': 939, 'Scene': 790, 'Description': 426, 'work_id': 902}

En este caso, nos interesa además revisar que no existen duplicados en la combinación ```<work_id, Act, Scene, Description>``` que identifica semánticamente a una escena (notar que cada fila está identificada por la columna ```id```).

In [21]:
duplicate_counts = df_chapters.duplicated(
    subset=["work_id", "Act", "Scene", "Description"]
).sum()
print(f"Duplicados: {duplicate_counts}")

Duplicados: 0


In [22]:
# Asegurarse que todo chapter pertenece a un work válido
df_merge = pd.merge(df_chapters, df_works,
                    left_on="work_id", right_on="id", how="left")
no_match_count = df_merge["id_y"].isna().sum()
print(f"Chapters sin Work: {no_match_count}")

Chapters sin Work: 0


In [23]:
# Genera ydata profiling report
generate_pandas_report(df_chapters, "chapters", DATA_REPORTS)

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

A priori no hay datos duplicados en capítulos en base a la ausencia de duplicados en la columna ```id``` asi como también en la ausencia de duplicados en las tuplas ```<work_id, Act, Scene, Description>```. 

* ✅ No hay repetidos en columna ```id```
* ✅ No hay repetidos en columna ```Title```
* ✅ No hay repetidos en columna ```LongTitle```

Por otro lado: 
* ✅ Todos los capítulos referencian a una obra (work) válida en ```df_works```.

#### 3.2.3. Paragraphs

Repitamos el mismo procedimiento ahora utilizando el dataframe ```df_paragraphs``` que contiene los datos de párrafos de las obras.

In [24]:
# Muestra TOP DEFAULT_TOP_ROWS_DISPLAY filas del dataframe
df_paragraphs.head(DEFAULT_TOP_ROWS_DISPLAY)

id  ParagraphNum                                          PlainText  \
0  630863             3  [Enter DUKE ORSINO, CURIO, and other Lords; Mu...   
1  630864             4  If music be the food of love, play on;\nGive m...   
2  630865            19                         Will you go hunt, my lord?   
3  630866            20                                       What, Curio?   
4  630867            21                                          The hart.   
5  630868            22  Why, so I do, the noblest that I have:\nO, whe...   
6  630869            30  So please my lord, I might not be admitted;\nB...   
7  630870            39  O, she that hath a heart of that fine frame\nT...   
8  630871            48                                           [Exeunt]   
9  630872            50              [Enter VIOLA, a Captain, and Sailors]   

   character_id  chapter_id  
0          1261       18704  
1           840       18704  
2           297       18704  
3           840       18704  
4           297       18704  
5           840       18704  
6          1221       18704  
7           840       18704  
8          1261       18704  
9          1261       18705

In [25]:
# Listado de tipos para las columnas
df_paragraphs.dtypes

id               int64
ParagraphNum     int64
PlainText       object
character_id     int64
chapter_id       int64
dtype: object

In [26]:
# Valores descriptivos de cada columna
df_paragraphs.describe()

id  ParagraphNum  character_id    chapter_id
count   35465.000000  35465.000000  35465.000000  35465.000000
mean   648595.000000   1602.761709    691.810743  19141.743268
std     10238.007985    972.461530    373.110791    274.769025
min    630863.000000      3.000000      1.000000  18704.000000
25%    639729.000000    788.000000    358.000000  18901.000000
50%    648595.000000   1554.000000    667.000000  19124.000000
75%    657461.000000   2341.000000    953.000000  19324.000000
max    666327.000000   4095.000000   1266.000000  19648.000000

In [27]:
# Conteo de missing values
empty_values = count_empty_values(df=df_paragraphs)
empty_values

id              0
ParagraphNum    0
PlainText       0
character_id    0
chapter_id      0
dtype: int64

In [28]:
# Revisamos ocurrencias duplicadas en cada columna del dataframe
duplicate_counts = {
    col: df_paragraphs[col].duplicated().sum() for col in df_paragraphs.columns
}
duplicate_counts

{'id': 0,
 'ParagraphNum': 31495,
 'PlainText': 1849,
 'character_id': 34245,
 'chapter_id': 34520}

In [29]:
duplicate_counts = df_paragraphs.duplicated(
    subset=["ParagraphNum", "PlainText", "character_id", "chapter_id"]
).sum()
print(f"Duplicados: {duplicate_counts}")

Duplicados: 0


In [30]:
# Asegurarse que todo paragraph pertenece a un chapter válido
df_merge = pd.merge(
    df_paragraphs, df_chapters, left_on="chapter_id", right_on="id", how="left"
)
no_match_count = df_merge["id_y"].isna().sum()
print(f"Párrafo con Chapter inexistente: {no_match_count}")

Párrafo con Chapter inexistente: 0


In [31]:
# Asegurarse que todo paragraph referencia a un character válido
df_merge = pd.merge(
    df_paragraphs, df_characters, left_on="character_id", right_on="id", how="left"
)
no_match_count = df_merge["id_y"].isna().sum()
print(f"Párrafo con Character inexistente: {no_match_count}")

Párrafo con Character inexistente: 0


Revisemos si existen párrafos asociados a más de un personaje lo cual no parecería muy lógico ya que en obras de teatro los párrafos son principalmente diálogos y los diálogos son de un solo personaje.

Como vemos en la consultra siguiente, no vemos duplicados a nivel de chapter_id, ParagraphNum, PlainText por lo que podemos concluir que cada combinación esta asociada a un único character_id.

**SPOILER ALERT**: Si hay diálogos que son repetidos por varios personajes en obras de teatro de Shakespeare pero a nivel de datos están registrados a nombre de un personaje muy peculiar. Seguir leyendo para descrubrirlo!!

In [32]:
grouped = df_paragraphs.groupby(['chapter_id', 'ParagraphNum', 'PlainText']).size().reset_index(name='count')
sorted_df = grouped.sort_values(by='count', ascending=False)
sorted_df

chapter_id  ParagraphNum  \
0           18704             3   
23645       19242          2658   
23639       19242          2651   
23640       19242          2652   
23641       19242          2653   
...           ...           ...   
11817       18986          2620   
11816       18986          2619   
11815       18986          2618   
11814       18986          2617   
35464       19648          3504   

                                               PlainText  count  
0      [Enter DUKE ORSINO, CURIO, and other Lords; Mu...      1  
23645             No, truly, but in friendly recompense.      1  
23639                                Do not you love me?      1  
23640                    Troth, no; no more than reason.      1  
23641  Why, then my cousin Margaret and Ursula\nAre m...      1  
...                                                  ...    ...  
11817  Say, gentle princess, would you not suppose\nY...      1  
11816            I cry you mercy, 'tis but Quid for Quo.      1  
11815                       Lady, wherefore talk you so?      1  
11814           Tush, women have been captivate ere now.      1  
35464                                           [Exeunt]      1  

[35465 rows x 4 columns]

In [33]:
grouped = df_paragraphs.groupby(['chapter_id', 'PlainText']).size().reset_index(name='count')
sorted_df = grouped.sort_values(by='count', ascending=False)
sorted_df

chapter_id                                          PlainText  count
240         18708                                             [Exit]      7
21981       19220                                             [Exit]      6
28341       19370                                             [Exit]      6
8777        18901                                 [Within]  Francis!      5
6618        18863                                             [Exit]      5
...           ...                                                ...    ...
11739       18986                                           [Exeunt]      1
11738       18986                     [Exeunt REIGNIER and MARGARET]      1
11737       18986        [Alarum. Excursions. Enter JOAN LA PUCELLE]      1
11736       18986  [Alarum. Enter SUFFOLK with MARGARET in his hand]      1
35106       19648                                           [Exeunt]      1

[35107 rows x 3 columns]

In [ ]:
# Genera ydata profiling report
generate_pandas_report(df_paragraphs, "paragraphs", DATA_REPORTS)

#### 3.2.4. Characters

Finalmente, repitamos el mismo procedimiento ahora utilizando el dataframe ```df_characters``` que contiene los datos de personajes de las obras.

In [34]:
# Muestra TOP DEFAULT_TOP_ROWS_DISPLAY filas del dataframe
df_characters.head(DEFAULT_TOP_ROWS_DISPLAY)

id           CharName             Abbrev Description
0   1   First Apparition   First Apparition         NaN
1   2      First Citizen      First Citizen         NaN
2   3  First Conspirator  First Conspirator         NaN
3   4    First Gentleman    First Gentleman         NaN
4   5         First Goth         First Goth         NaN
5   6     First Murderer     First Murderer         NaN
6   7     First Musician     First Musician         NaN
7   8     First Musician     First Musician         NaN
8   9      First Officer      First Officer         NaN
9  10       First Player              1Play         NaN

In [35]:
# Listado de tipos para las columnas
df_characters.dtypes

id              int64
CharName       object
Abbrev         object
Description    object
dtype: object

In [36]:
# Valores descriptivos de cada columna
df_characters.describe()

id
count  1266.000000
mean    633.500000
std     365.607029
min       1.000000
25%     317.250000
50%     633.500000
75%     949.750000
max    1266.000000

In [37]:
# Conteo de missing values
empty_values = count_empty_values(df=df_characters)
empty_values

id               0
CharName         0
Abbrev           5
Description    646
dtype: int64

In [38]:
# Revisamos ocurrencias duplicadas en cada columna del dataframe
duplicate_counts = {
    col: df_characters[col].duplicated().sum() for col in df_characters.columns
}
duplicate_counts

{'id': 0, 'CharName': 309, 'Abbrev': 302, 'Description': 799}

In [39]:
duplicate_counts = df_characters.duplicated(
    subset=["CharName", "Abbrev"]).sum()
print(f"Duplicados: {duplicate_counts}")

Duplicados: 270


In [40]:
duplicate_counts = df_characters.duplicated(
    subset=["id", "CharName", "Abbrev"]).sum()
print(f"Duplicados: {duplicate_counts}")

Duplicados: 0


Como podemos ver en las celdas anteriores tenemos dos caracterśticas interesantes en relación a los personajes:

1. La columna Description presenta una cantidad relevante de valores faltantes (NaN más especificamente)
2. Hay varios personajes repetidos que podrían tener un significado especial

Analicemos más en detalle esto agrupando por las columnas CharName y Abbrev a ver que podemos encontrar.

In [41]:
df_characters["count_duplicated"] = df_characters.groupby(["CharName"])[
    "CharName"
].transform("size")
df_duplicated = df_characters.drop_duplicates(subset=["CharName", "Abbrev"])
df_duplicated = df_duplicated.sort_values(
    by="count_duplicated", ascending=False)
df_duplicated.head(30)

id          CharName            Abbrev          Description  \
67      68               All               All                  NaN   
778    779         Messenger              Mess                  NaN   
768    769         Messenger         Messenger                  NaN   
85      86               All               ALL                  NaN   
772    773         Messenger         MESSENGER                  NaN   
1048  1049           Servant           Servant  servant to Diomedes   
1059  1060           Servant           SERVANT                  NaN   
680    681              Lord              Lord                  NaN   
675    676              Lord              LORD                  NaN   
848    849              Page              PAGE          to Falstaff   
20      21  Second Gentleman  Second Gentleman                  NaN   
847    848              Page              Page                  NaN   
434    435        First Lord        FIRST LORD                  NaN   
849    850              Page               Boy   a page to Gardiner   
432    433        First Lord        First Lord                  NaN   
3        4   First Gentleman   First Gentleman                  NaN   
516    517         Gentleman         Gentleman                  NaN   
215    216           Captain           Captain                  NaN   
218    219           Captain              Capt                  NaN   
518    519         Gentleman              Gent                  NaN   
179    180              Both              Both                  NaN   
1051  1052     First Servant     FIRST SERVANT                  NaN   
12      13     First Servant     First Servant                  NaN   
27      28    Second Servant    Second Servant                  NaN   
1053  1054    Second Servant    SECOND SERVANT                  NaN   
1002  1003       Second Lord       SECOND LORD                  NaN   
1000  1001       Second Lord       Second Lord                  NaN   
1        2     First Citizen     First Citizen                  NaN   
580    581            Herald               Her                  NaN   
14      15     First Soldier     First Soldier                  NaN   

      count_duplicated  
67                  23  
778                 23  
768                 23  
85                  23  
772                 23  
1048                21  
1059                21  
680                  9  
675                  9  
848                  8  
20                   8  
847                  8  
434                  8  
849                  8  
432                  8  
3                    8  
516                  7  
215                  7  
218                  7  
518                  7  
179                  7  
1051                 7  
12                   7  
27                   6  
1053                 6  
1002                 6  
1000                 6  
1                    6  
580                  5  
14                   5

In [42]:
paragraphs_all = df_paragraphs[(df_paragraphs["character_id"] == 68) | (df_paragraphs["character_id"] == 86)]
paragraphs_all_ext = pd.merge(paragraphs_all, df_chapters,
                    left_on="chapter_id", right_on="id", how="left")
paragraphs_all_ext = pd.merge(paragraphs_all_ext, df_works,
                    left_on="work_id", right_on="id", how="left")
paragraphs_all_ext[["PlainText", "character_id", "chapter_id", "Act", "Scene", "Description", "Title"]]

PlainText  character_id  \
0                                        Fly! not we.            68   
1                                    The gods forbid!            68   
2   [Speaking together] How now!\nHow now! do you ...            68   
3                              Content. 'Tis strange.            68   
4                               Good morrow, general.            68   
5                                      Alas, and woe!            68   
6                                     Most heavy day!            68   
7                                      A heavy sight!            68   
8                                          Dolabella!            68   
9   Fair is foul, and foul is fair:\nHover through...            86   
10  The weird sisters, hand in hand,\nPosters of t...            86   
11                                            So all.            86   
12                                    Well contented.            86   
13  Double, double toil and trouble;\nFire burn, a...            86   
14  Double, double toil and trouble;\nFire burn an...            86   
15  Double, double toil and trouble;\nFire burn an...            86   
16                             A deed without a name.            86   
17  Come, high or low;\nThyself and office deftly ...            86   
18                        Listen, but speak not to't.            86   
19                              Seek to know no more.            86   
20  Show his eyes, and grieve his heart;\nCome lik...            86   
21                            Hail, King of Scotland!            86   

    chapter_id  Act  Scene                                   Description  \
0        18767    3     11             Alexandria. CLEOPATRA’s palace.\n   
1        18771    4      2               Alexandria. CLEOPATRA’s palace.   
2        18772    4      3                  The same. Before the palace.   
3        18772    4      3                  The same. Before the palace.   
4        18773    4      4               The same. A room in the palace.   
5        18783    4     14                     The same. Another room.\n   
6        18783    4     14                     The same. Another room.\n   
7        18784    4     15                       The same. A monument.\n   
8        18785    5      1           Alexandria. OCTAVIUS CAESAR’s camp.   
9        19129    1      1                               A desert place.   
10       19131    1      3                          A heath near Forres.   
11       19138    2      3                                     The same.   
12       19138    2      3                                     The same.   
13       19146    4      1  A cavern. In the middle, a boiling cauldron.   
14       19146    4      1  A cavern. In the middle, a boiling cauldron.   
15       19146    4      1  A cavern. In the middle, a boiling cauldron.   
16       19146    4      1  A cavern. In the middle, a boiling cauldron.   
17       19146    4      1  A cavern. In the middle, a boiling cauldron.   
18       19146    4      1  A cavern. In the middle, a boiling cauldron.   
19       19146    4      1  A cavern. In the middle, a boiling cauldron.   
20       19146    4      1  A cavern. In the middle, a boiling cauldron.   
21       19156    5      8                    Another part of the field.   

                   Title  
0   Antony and Cleopatra  
1   Antony and Cleopatra  
2   Antony and Cleopatra  
3   Antony and Cleopatra  
4   Antony and Cleopatra  
5   Antony and Cleopatra  
6   Antony and Cleopatra  
7   Antony and Cleopatra  
8   Antony and Cleopatra  
9                Macbeth  
10               Macbeth  
11               Macbeth  
12               Macbeth  
13               Macbeth  
14               Macbeth  
15               Macbeth  
16               Macbeth  
17               Macbeth  
18               Macbeth  
19               Macbeth  
20               Macbeth  
21               Macbeth

In [43]:
# Soy curioso, que descripción tendrá y si aparece repetido Hamlet?
df_characters[df_characters["CharName"] == "Hamlet"] 

id CharName Abbrev                                        Description  \
558  559   Hamlet    Ham  son of the former king and nephew to the prese...   

     count_duplicated  
558                 1

De las consultas anterior podemos sacar mejores conclusiones acerca de los personajes repetidos:

1. Existen algunos personajes "especiales" con un significado que trasiende a una obra, más allá de un personaje conocido y con nombre y apellido como _Hamlet_. Este es el caso por ejemplo de: Servant, Lord, First Soldier, Second Lord. Estos son personajes genéricos que pueden aparecer en cualquier obra y podrían desviar algún tipo de análisis que queramos hacer ya que pueden estar presentes en múltiples obras con mismo o similar nombre, pero diferente id.
2. Vemos personajes que se llaman igual escritos de diferente forma como Second Lord con Abbrev en minúsculas y mayúsculas (1003 y 1001)
3. Hay dos personajes que se los más repetidos: All (y sus variantes) y Messenger (y sus variantes).

En resumen, si nos interesa posteriormente por ejemplo identificar los personajes con más párrafos dentro de una obra podemos contar los párrafos mediante un join entre las tablas Characters y Paragraph. Sin embargo si queremos llevar este análisis a la obra completa de Shakespeare, de repente sería bueno plantearnos: ¿Nos interesaría tratar a todos los _Messenger_ o _First Servant_ como uno mismo o diferentes? 

En función de la respuesta dependerá el tipo de procesamiento previo necesario a los datos.

In [ ]:
# Genera ydata profiling report
generate_pandas_report(df_characters, "characters", DATA_REPORTS)

## 4. Procesamiento de los Datos <a name="data-processing"></a>
[Volver al Inicio](#index)

In [44]:
# Creamos una nueva columna CleanText a partir de PlainText
df_paragraphs["CleanText"] = clean_text(df_paragraphs, "PlainText")

# Veamos la diferencia
df_paragraphs[["PlainText", "CleanText"]]

PlainText  \
0      [Enter DUKE ORSINO, CURIO, and other Lords; Mu...   
1      If music be the food of love, play on;\nGive m...   
2                             Will you go hunt, my lord?   
3                                           What, Curio?   
4                                              The hart.   
...                                                  ...   
35460  That she is living,\nWere it but told you, sho...   
35461  You gods, look down\nAnd from your sacred vial...   
35462  There's time enough for that;\nLest they desir...   
35463  O, peace, Paulina!\nThou shouldst a husband ta...   
35464                                           [Exeunt]   

                                               CleanText  
0       enter duke orsino  curio  and other lords  mu...  
1      if music be the food of love  play on  give me...  
2                             will you go hunt  my lord   
3                                           what  curio   
4                                              the hart   
...                                                  ...  
35460  that she is living  were it but told you  shou...  
35461  you gods  look down and from your sacred vials...  
35462  there s time enough for that  lest they desire...  
35463  o  peace  paulina  thou shouldst a husband tak...  
35464                                            exeunt   

[35465 rows x 2 columns]

In [45]:
# Convierte párrafos en listas "palabra1 palabra2 palabra3" -> ["palabra1", "palabra2", "palabra3"]
df_paragraphs["WordList"] = df_paragraphs["CleanText"].str.split()

# Nuevo dataframe: cada fila ya no es un párrafo, sino una sóla palabra
df_words = df_paragraphs.explode("WordList")

# Quitamos estas columnas redundantes
df_words.drop(columns=["CleanText", "PlainText"], inplace=True)

# Renombramos la columna WordList -> word
df_words.rename(columns={"WordList": "word"}, inplace=True)

# Verificar que el número de filas es mucho mayor
df_words

id  ParagraphNum  character_id  chapter_id     word
0      630863             3          1261       18704    enter
0      630863             3          1261       18704     duke
0      630863             3          1261       18704   orsino
0      630863             3          1261       18704    curio
0      630863             3          1261       18704      and
...       ...           ...           ...         ...      ...
35463  666326          3483           667       19648        d
35463  666326          3483           667       19648  hastily
35463  666326          3483           667       19648     lead
35463  666326          3483           667       19648     away
35464  666327          3504          1261       19648   exeunt

[909359 rows x 5 columns]

## 5. Análisis de los Datos <a name="data-analysis"></a>
[Volver al Inicio](#index)

En esta sección realizamos diferentes análisis sobre los datos, persiguiendo los siguientes objetivos (en su mayoría descritos en la letra del Laboratorio):

1. Analizar las obras de Shakespeare a través de los años e intentar identificar tendencias
2. Analizar frecuencias de palabras en la obra de Shakespeare para obtener una perspectiva diferente
3. Analizar los personajes con mayor cantidad de palabras asociadas en la obra de Shakespeare

También proponemos análisis adicionales que podríamos llevar adelante con más tiempo.

### 5.1. Obras de Shakespeare a través de los años

Para analizar las obras de Shakespeare a través de ños años alcanza con agrupar las filas de ```df_works``` por año y apartir de allí empear a analizar los resultados. Por ejemplo podemos ver la cantidad de obras producidas por año, o tomar ventanas de tiempo más grande como 5-10 años y analizar la producción desde allí. También puede ser interesante analizar como se distribuye la producción por género literario y ver como evolucionan estas proporciones a traves de los años.

Algunos de estos análisis son realizados a continuación con el objetivo de entender la obra de Shakespeare y detectar alguna posible tendencia. 

Un enfoque complementario, que tambien abordamos en esta sección es recurrir a la literatura. Cualquier autor tiene períodos en su vida que son fácilmente reconocibles en su obra y son ampliamente estudiados por expertos. En el caso de William Shakespeare de acuerdo a [10] se pueden apreciar cuatro períodos en su obra:

* In the Workshop (1589-1593)
* In the World (1594-1600)
* Out of the Depths (1601-1607)
* On the Heights (1608-1612)

En esta sección a su vez tomamos en cuenta esas referencias para revisar si hay diferencias notorias en los datos que acompañen la teoría.

In [46]:
# Definición de períodos
colors = px.colors.qualitative.Pastel2
shakespeare_born = 1564
shakespeare_periods = {
    1: {
        "name": "In the Workshop (1589-1593)",
        "from": 1589,
        "to": 1593,
        "color": colors[0],
    },
    2: {
        "name": "In the World (1594-1600)",
        "from": 1594,
        "to": 1600,
        "color": colors[1],
    },
    3: {
        "name": "Out of the Depths (1601-1607)",
        "from": 1601,
        "to": 1607,
        "color": colors[2],
    },
    4: {
        "name": "On the Heights (1608-1612)",
        "from": 1608,
        "to": 1612,
        "color": colors[3],
    },
}

In [48]:
# Primero agrupamos por Date y GenreType
works_per_year = df_works.groupby(["Date", "GenreType"]).size().unstack(fill_value=0)

# Creo una nueva columna con el Total Works por Date
works_per_year["Total"] = works_per_year.sum(axis=1)
works_per_year

GenreType  Comedy  History  Poem  Sonnet  Tragedy  Total
Date                                                    
1589            1        0     0       0        0      1
1590            0        2     0       0        0      2
1591            0        1     0       0        0      1
1592            0        1     0       0        0      1
1593            1        0     1       0        1      3
1594            2        0     1       0        1      4
1595            1        1     0       0        0      2
1596            1        1     0       0        0      2
1597            0        2     0       0        0      2
1598            1        1     1       0        0      3
1599            2        0     0       0        1      3
1600            1        0     0       0        1      2
1601            0        0     1       0        1      2
1602            1        0     0       0        0      1
1604            1        0     0       0        1      2
1605            0        0     0       0        2      2
1606            0        0     0       0        1      1
1607            0        0     0       0        2      2
1608            0        1     0       0        0      1
1609            0        1     1       1        0      3
1610            1        0     0       0        0      1
1611            1        0     0       0        0      1
1612            0        1     0       0        0      1

Una primera visualización para ver la cantidad de obras por año y por género, dónde cada bloque indica una obra.

In [49]:
works_per_year_without_total = works_per_year.drop(columns=["Total"])
fig, ax = plt.subplots( figsize=(15, 6))
bottom= 0

for genre, years in works_per_year_without_total.items():

    ax.bar(years.index, label=genre, height=years.values,bottom=bottom)
    bottom = bottom + years.values
plt.xticks(np.arange(1589, 1613,1))
plt.title("Obras de Shakespeare por año y género")
plt.ylabel("Cantidad de obras")
plt.xlabel("Año")
plt.legend()

In [50]:
# Obras de Shakespeare por año
fig = px.bar(works_per_year, x=works_per_year.index, y="Total", )
fig.update_xaxes(tickmode="array", tickvals=works_per_year.index)
fig.update_layout(
    height=400,
    width=1200,
    title_text="Obras de William Shakespeare (por año)",
    xaxis_title="Año",
    yaxis_title="Obras Producidas (anual)",
)
fig.show()

# Guardar imagen: Descomentar para guardar
#fig.write_image(os.path.join(FIGURES_FOLDER, "obras_por_año_1.png"))

Veamos como queda la anterior visualización cuando marcamos explícitamente los períodos en la vida de Shakespeare, descritos por estudiosos de la obra del poeta.

In [51]:
fig = px.bar(works_per_year, x=works_per_year.index, y="Total")
fig.update_xaxes(tickmode="array", tickvals=works_per_year.index)

for idx, period in shakespeare_periods.items():
    fig.add_vrect(
        x0=period["from"] - 0.5,
        x1=period["to"] + 0.5,
        annotation_text=period["name"],
        annotation_position="top",
        annotation_font_color="blue",
        annotation=dict(font_size=15, font_family="Arial"),
        fillcolor=period["color"],
        opacity=0.5,
        line_width=0,
    )
fig.update_layout(
    height=400,
    width=1100,
    title_text="Obras de William Shakespeare (por año)",
    xaxis_title="Año",
    yaxis_title="Obras Producidas (anual)",
)
fig.show()

# Guardar imagen. Descomentar para guardar
#fig.write_image(os.path.join(FIGURES_FOLDER, "obras_por_año_2.png"))

Puede resultar un poco más intuitivo para entender la evolución del escritor asi como explicar posibles tendencias, analizar la evolución a lo largo de la vida de Shakespeare, comparando contra su edad en lugar de a través de los años.

In [52]:
df_works["Age"] = df_works["Date"] - shakespeare_born

# Primero agrupamos por Date y GenreType
works_per_age = df_works.groupby(["Age", "GenreType"]).size().unstack(fill_value=0)

# Creo una nueva columna con el Total Works por Date
works_per_age["Total"] = works_per_age.sum(axis=1)

fig = px.bar(works_per_age, x=works_per_age.index, y="Total")
fig.update_xaxes(tickmode="array", tickvals=works_per_age.index)
for idx, period in shakespeare_periods.items():
    fig.add_vrect(
        x0=period["from"] - shakespeare_born - 0.5,
        x1=period["to"] - shakespeare_born + 0.5,
        annotation_text=period["name"],
        annotation_position="top",
        annotation_font_color="blue",
        annotation=dict(font_size=15, font_family="Arial"),
        fillcolor=period["color"],
        opacity=0.5,
        line_width=0,
    )

fig.update_layout(
    height=400,
    width=1100,
    title_text="Obras de William Shakespeare (por edad)",
    xaxis_title="Edad (años)",
    yaxis_title="Obras Producidas (anual)",
)
fig.show()

# Guardar imagen. Descomentar para guardar
#fig.write_image(os.path.join(FIGURES_FOLDER, "obras_por_edad_1.png"))

In [53]:
works_per_year_cumulative = pd.DataFrame(index=works_per_age.index)
for col in works_per_age.columns:
    works_per_year_cumulative[col] = works_per_age[col].cumsum()

plot = go.Figure()

for col in works_per_year_cumulative.columns:
    if col != "Total":
        plot.add_trace(
            go.Scatter(
                name=col,
                x=works_per_year_cumulative.index,
                y=works_per_year_cumulative[col],
                stackgroup="one",
            )
        )

plot.update_layout(
    height=600,
    width=1100,
    title_text="Obras de William Shakespeare",
    xaxis_title="Edad (años)",
    yaxis_title="Acumulado de Obras",
)
plot.show()

# Guardar imagen. Descomentar para guardar
#plot.write_image(os.path.join(FIGURES_FOLDER, "obras_por_edad_2.png"))

In [54]:
from plotly.subplots import make_subplots

fig = make_subplots(rows=5, cols=1)

for ix, col in enumerate(works_per_year_cumulative.columns):
    if col != "Total":
        fig.append_trace(
            go.Scatter(
                x=works_per_year_cumulative.index,
                y=works_per_year_cumulative[col],
                name=col,
                line_shape="hv",
            ),
            row=ix + 1,
            col=1,
        )


fig.update_layout(height=800, width=1100, title_text="Obras de William Shakespeare")
# fig.update_xaxes(tickmode='array',tickvals=works_per_year.index)

for idx, period in shakespeare_periods.items():
    fig.add_vrect(
        x0=period["from"] - shakespeare_born - 0.5,
        x1=period["to"] - shakespeare_born + 0.5,
        annotation_text="",
        annotation_position="top",
        annotation_font_color="blue",
        annotation=dict(font_size=15, font_family="Arial"),
        fillcolor=period["color"],
        opacity=0.5,
        line_width=0,
    )

fig.show()

# Guardar imagen. Descomentar para guardar
# fig.write_image(os.path.join(FIGURES_FOLDER, "obras_por_edad_3.png"))

### 5.2. Conteo de palabras frecuentes

Mediante el análisis de palabras frecuentes podemos obtener una visión diferente sobre el usuo del vocabulario en las obras de William Shakespeare. Para ello lo que vamos a hacer es calcular las frecuencias de palabras y para ello es necesario primero realizar ciertos pre-procesamientos al texto que nos permitan calcualar estas frecuencias de forma acertada:

1. Normalizar las palabras: Llevar a minúsculas (lowercase) y remover espacios innecesarios (strip). Notar que esto ya lo hicimos en ```CleanText```
2. Separar el texto en palabras: Separar el texto en palabras o tokens para contar posteriromente la frecuencia. Notar que esto ya lo hicimos en ```WordList```.
3. Remover stopwords: Algunas palabras como "and" o "or" son ampliamente utilizadas en el lenguaje pero no aportan mayor valor a nuestro análizis. De considerarlas seguramente se roben el protagonismo de cualquier análisis. Por ello primero vamos a remover stopwords de la lista de palabras ```WordList```, para quedarnos solamente con las palabras más relevantes del idioma.

Notar que el paso 2 lo estamos realizando de forma sencilla mediane la función [split()](https://www.w3schools.com/python/ref_string_split.asp) de strings que por defecto utiliza como separador un whitespace ```" "```. 

```python
df_paragraphs["WordList"] = df_paragraphs["CleanText"].str.split()
```

En un trabajo más profundo, podriamos reemplazar este método de tokenización por métodos más complejos y precisos como el [Tokenizer](https://spacy.io/api/tokenizer) de spacy que contempla otro tipo de separadores naturales en el idioma ingles como ```,?-;:```, etc.

Para implementar el punto 3 vamos a utilizar la librería de NLP [spacy](https://spacy.io/) que implementa métodos sencillos para remvoer stopwords de un texto. Además, nos evita la tediosa tarea de conseguir las stopwors para un idioma en particular (en este caso inglés), lista de tokens que puede ser bastante larga.

In [55]:
# Veamos las stopwords de Spacy para Inglés
from spacy.lang.en import stop_words

stop_words = stop_words.STOP_WORDS
print(
    f"Este modelo contiene {len(stop_words)} stop words. Las primeras 10 son: {list(stop_words)[:15]}"
)

Este modelo contiene 326 stop words. Las primeras 10 son: ['when', 'something', 'one', '’ll', 'thru', '‘d', 'amongst', 'per', 'been', 'then', 'on', 'formerly', 'no', 'another', 'would']


A continuación limpiamos las palabras usando el modelo ```en_core_web_sm```. Spacy tiene varios modelos en función del idioma. Para esta tarea vamos a utilizar el modelo más sencillo de ingles. Por más información sobre los modelos de Spacy ver [Trained Models & Pipelines](https://spacy.io/models).

In [56]:
def chunks(lst: List[Any], n: int):
    """Funcion auxiliar para generar n-sized chunks desde lst."""
    for i in range(0, len(lst), n):
        yield lst[i : i + n]


# Cargamos modelo spacy (modelo liviano para ingles)
nlp = spacy.load("en_core_web_sm", disable=["ner", "tagger", "parser", "textcat"])

# Custom stop words
customize_stop_words = []
for w in customize_stop_words:
    nlp.vocab[w].is_stop = True

words = df_words["word"].values

# Filtro las words de todo el texto manteniendo las que no son stopwords
clean_words = []
for words_batch in tqdm.tqdm(chunks(words, 1000)):
    text = " ".join(words_batch)
    clean_words += [token.lemma_ for token in nlp(text) if not token.is_stop]

# Veamos las primeras
print(f"Primeras 10 palabras limpias: {clean_words[:10]}")

0it [00:00, ?it/s]/home/eviotti/.pyenv/versions/3.12.0/envs/intro-cd/lib/python3.12/site-packages/spacy/pipeline/lemmatizer.py:211: UserWarning:

[W108] The rule-based lemmatizer did not find POS annotation for one or more tokens. Check that your pipeline includes components that assign token.pos, typically 'tagger'+'attribute_ruler' or 'morphologizer'.

910it [00:42, 21.56it/s]

Primeras 10 palabras limpias: ['enter', 'duke', 'orsino', 'curio', 'lords', 'musicians', 'attending', 'music', 'food', 'love']


In [58]:
# Construimos wordcloud a partir de las frecuencias de todas las obras sin distinción
words = " ".join(clean_words)
wc = WordCloud(
    background_color="white",
    max_words=2000,
    # mask=alice_mask,
    contour_width=3,
    contour_color="steelblue",
)

# generate word cloud
wc.generate(words)

# store to file. Descomentar para guardar
# wc.to_file(os.path.join("assets", "snapshoots", "wordcloud.png"))

# show
plt.imshow(wc, interpolation="bilinear")
plt.axis("off")
plt.show()
# Save image. Descomentar para guardar
# plt.savefig(os.path.join(FIGURES_FOLDER, "wordcloud.png"))

/tmp/ipykernel_45009/2353396249.py:20: UserWarning:

FigureCanvasAgg is non-interactive, and thus cannot be shown



En el WordCloud creado a partir de todas la palabras de la obra y sin los stop_words, aparecen palabras pero que en realidad son letras únicas cómo lo son "d", "s" y "o". Luego de una breve investiagación esto puede deberse a que contracciones utilizadas por Shakespeare que no son utilizadas hoy en día, por ejemplo "o'er" siendo esta una contracción de over. 

Continuando con este análisis, se puede realizar la visualización de las palabras mas comunes en forma de wordcloud para cada género. Para poder realizar este análisis es necesario poder vincular las palabras al género que fueron utilizadas, debido a las relaciones que tinene las tablas ese necesario realizar diferentes merges para poder obtener la relación buscada. Cómo se mostró en la sección 3, se debe vincular el párrafo a capítulos para obtener los párrafos vinculados a géneros y luego unir las palabras con su párrafo de procedencia.

In [59]:
df_paragraphs_with_chapter= pd.merge(df_paragraphs[["id","ParagraphNum", "chapter_id"]], df_chapters[["id", "work_id"]], left_on="chapter_id", right_on="id")
df_paragraphs_with_genre = pd.merge(df_paragraphs_with_chapter, df_works[["id", "GenreType"]], left_on="work_id", right_on="id")
df_words_with_genre = pd.merge(df_words, df_paragraphs_with_genre[["id_x","ParagraphNum", "GenreType"]], left_on="id", right_on="id_x", how="left")
df_words_with_genre 


id  ParagraphNum_x  character_id  chapter_id     word    id_x  \
0       630863               3          1261       18704    enter  630863   
1       630863               3          1261       18704     duke  630863   
2       630863               3          1261       18704   orsino  630863   
3       630863               3          1261       18704    curio  630863   
4       630863               3          1261       18704      and  630863   
...        ...             ...           ...         ...      ...     ...   
909354  666326            3483           667       19648        d  666326   
909355  666326            3483           667       19648  hastily  666326   
909356  666326            3483           667       19648     lead  666326   
909357  666326            3483           667       19648     away  666326   
909358  666327            3504          1261       19648   exeunt  666327   

        ParagraphNum_y GenreType  
0                    3    Comedy  
1                    3    Comedy  
2                    3    Comedy  
3                    3    Comedy  
4                    3    Comedy  
...                ...       ...  
909354            3483    Comedy  
909355            3483    Comedy  
909356            3483    Comedy  
909357            3483    Comedy  
909358            3504    Comedy  

[909359 rows x 8 columns]

Generamos una visualización con el conjunto de wordcloud por género.

In [60]:
plt.figure(figsize=(15, 6))
for index, i in enumerate(["Tragedy", "Comedy" , "History", "Sonnet","Poem"]):
    print(i, index)
    words = df_words_with_genre.groupby("GenreType")["word"].value_counts()[i].index
    words = " ".join(words)
    wc = WordCloud(
        background_color="white",
        max_words=2000,
        # mask=alice_mask,
        contour_width=3,
        contour_color="steelblue",
    )
    wc.generate(words)
    plt.subplot(2, 3, index+1)
    plt.title("WordCloud de "+i)
    plt.imshow(wc, interpolation="bilinear")
    plt.axis("off")
plt.tight_layout()

# Save image. Descomentar para guardar
#plt.savefig(os.path.join(FIGURES_FOLDER, "wordcloud_by_genre.png"))

Tragedy 0
Comedy 1
History 2
Sonnet 3
Poem 4


De esta visualización se logran ver algunas tendencias, por ejemplo en POem y Sonnet aparece la palabra "love "destacada mientras que en el resto de los géneros no.

### 5.3. Personajes con más cantidad de palabras

In [61]:
# Agregamos el nombre de los personajes
# TODO: des-comentar luego de cargar df_characters
df_words_char = pd.merge(
    df_words, df_characters[["id", "CharName"]], left_on="character_id", right_on="id"
)
df_words_char

id_x  ParagraphNum  character_id  chapter_id     word  id_y  \
0       630863             3          1261       18704    enter  1261   
1       630863             3          1261       18704     duke  1261   
2       630863             3          1261       18704   orsino  1261   
3       630863             3          1261       18704    curio  1261   
4       630863             3          1261       18704      and  1261   
...        ...           ...           ...         ...      ...   ...   
909354  666326          3483           667       19648        d   667   
909355  666326          3483           667       19648  hastily   667   
909356  666326          3483           667       19648     lead   667   
909357  666326          3483           667       19648     away   667   
909358  666327          3504          1261       19648   exeunt  1261   

                  CharName  
0       (stage directions)  
1       (stage directions)  
2       (stage directions)  
3       (stage directions)  
4       (stage directions)  
...                    ...  
909354             Leontes  
909355             Leontes  
909356             Leontes  
909357             Leontes  
909358  (stage directions)  

[909359 rows x 7 columns]

In [62]:
# TODO:
# - des-comentar luego de hacer el merge
# - Encuentra algún problema en los resultados?

words_per_character = (
    df_words_char.groupby("CharName")["word"].count().sort_values(ascending=False)
)
words_per_character

CharName
Poet                  50762
(stage directions)    16443
Henry V               15428
Falstaff              14906
Hamlet                12291
                      ...  
Philip                    3
Nicholas                  2
Joseph                    2
Taurus                    2
Thieves                   1
Name: word, Length: 925, dtype: int64

In [64]:
# Ejemplo: 10 personajes con más palabras
plt.figure(figsize=(8, 6))
char_show = words_per_character[:10]
char_show = char_show.sort_values()
plt.barh(char_show.index, char_show.values)
_ = plt.xticks(rotation=90)
plt.ylabel("Personaje")
plt.xlabel("Cantidad de palabras")
plt.title("Cantidad de palabras por personaje")
plt.show()

# Save image. Descomentar para guardar
#plt.savefig(os.path.join(FIGURES_FOLDER, "words_per_character.png"))


/tmp/ipykernel_45009/1820703871.py:10: UserWarning:

FigureCanvasAgg is non-interactive, and thus cannot be shown



In [65]:
words_per_character_df = pd.DataFrame(words_per_character).reset_index()
romeo = words_per_character_df[words_per_character_df['CharName'] == 'Romeo']
juliet = words_per_character_df[words_per_character_df['CharName'] == 'Juliet']

print(f"Las posiciones en el ranking para Romeo={romeo.index[0]} con {romeo['word'].sum()} palabras y para Julieta={juliet.index[0]} con {juliet['word'].sum()} palabras.")

Las posiciones en el ranking para Romeo=31 con 4873 palabras y para Julieta=38 con 4499 palabras.


Las celdas anteriors nos permiten identificar algunas cosas bien interesantes:

1. En el TOP 3 de personajes con mayor cantidad de palabras asociadas se destacan: Poet y (stage directions)
2. Los personajes particualres con mayor cantidad de palabras asociadas son: Henry V, Falstaff, Hamlet y Dule of Gloucester. Lejos quedaron del TOP 10 la famosa pareja de Romeo y Julieta (quienes por cierto manejan una cantidad similar de palabras).

Analizando más en detalle estos resultados:

1. Poet representa al narrador en los poemas escritos por Shakespeare
2. Stage Directions de acuerdo a [shakespearestagedirections.coe.edu](https://shakespearestagedirections.coe.edu/#:~:text=Stage%20directions%20are%20where%20the,or%20directors%20about%20that%20information.) son indicaciones particulares escritas por Shakespeare para facilitar la labor de los actores que interpreten un personaje, así como facilitar al lector visualizar la escena que esta ocurriendo. Por ello es de esperarse la aparición de varias acotaciones de este estilo y en particualr como en el dataset las Stage Directions son asociadas a un personaje "Stage Directions" es de esperar ver los resultados anteriores.

**Mejoras**

A partir del análisis anterior se nos ocurren diferentes mejoras para extender o complementar el análisis:

1. Remover Poet y (stage directions) del TOP ahora que sabemos que son dos categorias especiales de personajes.
2. Cambiar la forma en la que contamos las palabras por personaje ya que estamos contando stopwords que quizas no sean de tanto interés y 
además personajes de obras teatrales y comedias naturalmente tendrán mayor cantidad de palabras ya que son obras más extensas. Por ello 
proponemos normalizar los valores considerando la cantidad de palabras que tiene la obra en la que aparece el personaje.

In [ ]:

# 10 personajes con más parrafps retirando Poet y Stage Directions
plt.figure(figsize=(8, 6))

char_show = words_per_character[:12]
char_show = char_show[2:] # Sacamos poet y Stage Sirections
char_show = char_show.sort_values()
plt.barh(char_show.index, char_show.values)
_ = plt.xticks(rotation=90)
plt.ylabel("Personaje")
plt.xlabel("Cantidad de palabras")
plt.title("Cantidad de palabras por personaje")
plt.savefig(os.path.join(FIGURES_FOLDER, "words_per_character_modified.png"))


Podemos también analizar la cantidad de párrafos por personaje

In [ ]:
df_characters_with_char = pd.merge(df_paragraphs, df_characters, left_on="character_id", right_on="id", how="left")
top_characters_by_paragraph = df_characters_with_char.groupby("CharName")["CharName"].count().sort_values(ascending=False)

In [ ]:
# Ejemplo: 10 personajes con más palabras
plt.figure(figsize=(8, 6))
char_show = top_characters_by_paragraph[:10]
char_show = char_show.sort_values()
plt.barh(char_show.index, char_show.values)
_ = plt.xticks(rotation=90)
plt.ylabel("Personaje")
plt.xlabel("Cantidad de párrafos")
plt.title("Cantidad de párrafos por personaje")


In [ ]:
# Ejemplo: 10 personajes con más palabras
char_show = words_per_character[:12]
char_show = char_show[2:] # Sacamos poet y Stage Sirections
char_show = char_show.sort_values()
plt.barh(char_show.index, char_show.values)
_ = plt.xticks(rotation=90)
plt.ylabel("Personaje")
plt.xlabel("Cantidad de párrafos")
plt.title("Cantidad de párrafos por personaje")
plt.savefig(os.path.join(FIGURES_FOLDER, "paragraph_per_character.png"))



### 5.4. Preguntas Adicionales sobre los Datos

Ver sección **Análisis Complementarios** en [Laboratorio_1_informe.pdf](Laboratorio_1_informe.pdf).

## 6. Conclusiones <a name="conclusions"></a>
[Volver al Inicio](#index)

A partir de los análisis realizados al conjunto de datos con obras de William Shakespeare, podemos formular las siguientes conclusiones a partir de sus obras allí presente y al conjunto de datos en sí mismo.

* El autor comenzó a escribir a la tardía edad de 25 años si se toma en cuenta que muere a los 52 años. Una vida que de todas formas supera la expectativa de vida de un ciudadano londinense del siglo XVI-XVII.
* Pese a esto su obra fue muy prolífera con 43 obras entre Tragedias, Comedias, Poemas, Sonetos y Dramas Históricos.
* De toda su vida como escritor, su período más fructífero fue entre los años 1594 - 1600 en el período que se lo conoce como “In The World” por quienes estudian la obra del autor.
* Llama la atención el año 1603 en donde no vemos obras publicadas, lo cual contradice la tendencia histórica en su producción de obras.
* Su vida como escritor se encuentra claramente marcada por las comedias, 14 obras escritas (33 % de su obra), le siguen los dramas históricos con 12 obras (28% de su obra), y las tragedias con 11 obras escritas (26%). En su mayoría obras de teatro convirtiéndo a Shakespeare en un dramaturgo por excelencia.
* Dentro de la obra de Shakespeare, se utiliza un lenguaje un inglés diferente al actual y esto se ve claramente reflejado en las palabras más utilizadas por el escritor. 
* Los personajes que mayor diálogo tienen entre las obras del escritor son Henry V, Falstaff y Hamlet. Lejos de este ranking quedan los conocidos personajes Romeo y Julieta.

Por otro lado, analizando el conjunto de datos y su calidad, podemos concluir que:
* La calidad de los datos presentes es muy buena, mostrando una ausencia de problemas de calidad como datos faltantes, salvo por las columnas Abbrev y Description en la tabla characters en donde se ve una buena presencia de valores faltantes. 
* No se ven datos duplicados o incoherentes a simple vista. Esto a su vez tiene mucho sentido, tomando en cuenta de que se trata de un conjunto de datos que posiblemente se haya realizado de forma exhaustiva a partir de la obra de William Shakespeare.


## 7. Referencias <a name="references"></a>
[Volver al Inicio](#index)

1. [Base de Datos Shakespeare](https://relational-data.org/dataset/Shakespeare)
2. [Laboratorio 1](https://gitlab.fing.edu.uy/maestria-cdaa/intro-cd/)
3. [Referencia 1](www.google.com)
4. [Towards Data Science Exploratory Data Analisys](https://towardsdatascience.com/exploratory-data-analysis-8fc1cb20fd15)
5. [Data Camp: Exploratory Data Analysis in Python](https://www.datacamp.com/courses/exploratory-data-analysis-in-python)
6. [Wikipedia William Shakespeare](https://en.wikipedia.org/wiki/William_Shakespeare)
7. [Python Wordcloud](https://pypi.org/project/wordcloud/)
8. [Plotly Vertical Lines](https://plotly.com/python/horizontal-vertical-shapes/)
9. [Plotly Colors](https://plotly.com/python/discrete-color/)
10. [Four Periods of Shakespeare's Dramatic and Poetic Career](https://moirabaricollegeonline.co.in/attendence/classnotes/files/1589611082.docx#:~:text=Although%20the%20precise%20date%20of,the%20Fourth%20Period%20from%201608.)